precompiles/vesting/approve.go

```diff
@@ -0,0 +1,63 @@
+package vesting
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/authz"
+	authzkeeper "github.com/cosmos/cosmos-sdk/x/authz/keeper"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+	cmn "github.com/evmos/evmos/v14/precompiles/common"
+	vestingtypes "github.com/evmos/evmos/v14/x/vesting/types"
+)
+
+var (
+	// FundVestingAccountMsg defines the authorization type for MsgFundVestingAccount
+	FundVestingAccountMsg = sdk.MsgTypeURL(&vestingtypes.MsgFundVestingAccount{})
+	// UpdateVestingFunderMsg defines the authorization type for MsgUpdateVestingFunder
+	UpdateVestingFunderMsg = sdk.MsgTypeURL(&vestingtypes.MsgUpdateVestingFunder{})
+	// ClawbackMsg defines the authorization type for MsgDelegate
+	ClawbackMsg = sdk.MsgTypeURL(&vestingtypes.MsgClawback{})
+)
+
+// Approve is the precompile function for approving vesting transactions with a generic grant.
+func (p Precompile) Approve(
+	ctx sdk.Context,
+	origin common.Address,
+	_ vm.StateDB,
+	method *abi.Method,
+	args []interface{},
+) ([]byte, error) {
+	grantee, typeURL, err := checkApprovalArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	switch typeURL {
+	case FundVestingAccountMsg, ClawbackMsg, UpdateVestingFunderMsg:
+		if err := CreateGenericAuthz(ctx, p.AuthzKeeper, grantee, origin, typeURL); err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf(cmn.ErrInvalidMsgType, "vesting", typeURL)
+	}
+
+	// TODO: Add event emitting maybe ?
+
+	return method.Outputs.Pack(true)
+}
+
+// CreateGenericAuthz creates a generic authorization grant.
+func CreateGenericAuthz(
+	ctx sdk.Context,
+	authzKeeper authzkeeper.Keeper,
+	grantee, granter common.Address,
+	msg string,
+) error {
+	genericAuthorization := authz.GenericAuthorization{Msg: msg}
```

Comment:
this is fine for fund and clawback but the UpdateVestingFunder should definitely have a custom authorization otherwise there could be security implications where a vulnerable smart contract can change the funder and then clawback